### DISCUSSION TITLE
How do not push changes or cancel a pipeline if updating dependencies failed with "Artifact update problem"?

### DISCUSSION THREAD

**Discussion by simon-liebehenschel** [renovatebot/renovate#18989] [Request Help]
NOTE: This question is not about solving the "Artifact update problem" error itself. The question is about how Renovate behaves when such an error happens.

## Current behavior

1. Renovate fails to update dependencies. This is expected sometimes because sometimes Python packages have conflicting versions.
2. Renovate pushes changes (pushes "pyproject.toml", but not "poetry.lock").
3. A CI/CD pipeline is started.

## Why the current behavior is bad

If dependencies locking failed I do not need to run a CI/CD pipeline. Basically, I even do not need the pushed files.

## Desired behavior

1. Renovate fails to update dependencies.
2. Renovate does not push changes OR it pushes changes with a [skipped pipeline](https://docs.gitlab.com/ee/ci/pipelines/index.html#skip-a-pipeline).

How can I achieve this behavior? Renovate sometimes fails with `poetry lock` (which is expected sometimes if packages versions have conflicts) and Renovate wastes my CI/CD minutes by useless push and a pipeline.

Is there any built-in support for to do not run a pipeline or cancel it if an artifact update problem happens?
I would like to have an option `PUSH_CHANGES_IF_UPDATING_DEPENDENCIE_FAILED: bool` or `RUN_CI_PIPELINE_IF_UPDATING_DEPENDENCIES_FAILED: bool`. Was something like this discussed or implemented?

## Versions
- GitLab SaaS
- `renovate/renovate:34.24` Docker image

I can share my Renovate config, but I think it is unrelated to my question.

## Workaround

I can update GitLab CI/CD rules to do not trigger a pipeline if a `poetry.lock` file was not changed. This way I will avoid useless pipelines.

**Comment by rarkins** [comment]
So if there is an artifacts error, you'd prefer that the branch is never pushed? Forever?

**Reply by viceice** [reply]
I think the request is to push but set the skip CI option.

https://docs.gitlab.com/ee/user/project/push_options.html#push-options-for-gitlab-cicd

**Reply by simon-liebehenschel** [reply]
It is fine if changes will be pushed (e.g. "pyproject.toml", but not "poetry.lock"). It is fine that a branch will be created.

Two actions **must not** happen when Renovate already knows that updating dependencies failed:
1. CI/CD pipeline [must not be triggered](https://docs.gitlab.com/ee/user/project/push_options.html#push-options-for-gitlab-cicd) or, if a pipeline was started, then the pipeline must be cancelled by Renovate.
   I think this depends on what is easier to implement on a Renovate side. E.g. if cancelling is easier to implement, then Renovate can cancel a running pipeline.
2. Renovate must not call `Merge request was scheduled to merge after pipeline succeeds` or, if called, undo this option. Because (obviously) tests will fail since dependencies update failed.

To summarize one more time, here is a workflow: 

![renovate drawio](https://user-images.githubusercontent.com/70140627/202842805-723557bf-19bb-4d07-b0d2-f0114d10f653.svg)

**Reply by viceice** [reply]
cancel a pipeline isn't working, because we don't know when it's starting.

so the only solution for gitlab is to use the skip CI git push option.

**Reply by simon-liebehenschel** [reply]
Calling `git push -o ci.skip` looks simple to implement. Also, it looks logically to skip a CI pipeline when dependencies updating failed. This behavior can be optional (opt-in).
Should I transfer this discussion into a feature request? 

PS: I am sorry, but I am a Python developer, so I am very limited in ways how I can contribute here.

**Reply by philippe-granet** [reply]
Waiting for a solution, I use this patch (use at your own risk) :

````bash
sed -i '/if (config\.commitBody) {/a\
        if (config.artifactErrors?.length && !config.commitBody.includes('\''[skip ci]'\'') && !config.commitBody.includes('\''[ci skip]'\'') ) {\
            config.commitBody = `[skip ci] 🚨 Artifacts errors 🚨\\n\\n${config.commitBody}\\n\\n`;\
        }' \
          /usr/local/renovate/dist/workers/repository/update/branch/index.js
```
